### PR TITLE
KTOR-400 Download progress observer

### PR DESCRIPTION
**Pull request by rsinukov**
**Subsystem**
Client

**Motivation**
[Client upload/download progress observer/handler/interceptor](https://youtrack.jetbrains.com/issue/KTOR-400)

Second part of the feature with download progress.

API:
```kotlin
client.post("$TEST_SERVER/content/echo") {
  body = Body()
  downloadProgress { sent, total -> ... }
}
```

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpStatement.kt [hunk 1]
```diff
@@ -39,8 +43,14 @@ public class HttpStatement(
      *
      * Please note: the [response] instance will be canceled and shouldn't be passed outside of [block].
      */
-    public suspend fun <T> execute(block: suspend (response: HttpResponse) -> T): T {
-        val response: HttpResponse = executeUnsafe()
+    public suspend fun <T> execute(
+        listener: ProgressListener? = null,
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpStatement.kt:47]: I'm not sure that we should introduce such an API call for the specific feature. Could we make an extension point in the request builder instead?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpStatement.kt [hunk 2]
```diff
@@ -79,13 +89,23 @@ public class HttpStatement(
...
+
     /**
      * Executes this statement and run the [block] with a [HttpClient.responsePipeline] execution result.
      *
      * Note that T can be a streamed type such as [ByteReadChannel].
      */
-    public suspend inline fun <reified T, R> receive(crossinline block: suspend (response: T) -> R): R {
-        val response: HttpResponse = executeUnsafe()
+    public suspend inline fun <reified T, R> receive(
+        noinline listener: ProgressListener? = null,
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/statement/HttpStatement.kt:102]: same here

File: ktor-client/ktor-client-core/common/src/io/ktor/client/utils/ByteChannelUtils.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+
+internal fun ByteReadChannel.observable(
+    context: CoroutineContext,
+    contentLength: Long?,
+    listener: ProgressListener
+) = GlobalScope.writer(context, autoFlush = true) {
+    val total = contentLength ?: -1
+    val byteArray = ByteArray(8 * 1024)
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/utils/ByteChannelUtils.kt:18]: Consider using pool here

Comment (e5l) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/utils/ByteChannelUtils.kt:18]: I would also extract constants

**Review by e5l — CHANGES_REQUESTED** [review]

File: ktor-client/ktor-client-core/common/src/io/ktor/client/features/BodyProgress.kt [hunk 4]
```diff
@@ -52,3 +63,14 @@ public class BodyProgress internal constructor() {
         }
     }
 }
+
+@SharedImmutable
+private val DownloadProgressListenerAttributeKey = AttributeKey<ProgressListener>("UploadProgressListenerAttributeKey")
+
+public fun HttpRequestBuilder.downloadProgress(listener: ProgressListener?) {
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/features/BodyProgress.kt:70]: missing KDoc for the public method

**Review by e5l — APPROVED** [review]
Please add KDocs for the public methods

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/ObservableBodyTest.kt [hunk 5]
```diff
@@ -96,4 +98,114 @@ class ObservableBodyTest : ClientLoader() {
...
+            }
+        }
+
+        test { client ->
+            invokedCount = 0
+            val listener: ProgressListener = { _, _ -> invokedCount++ }
+
+            client.get<HttpStatement>("$TEST_SERVER/json/users-long") {
+                contentType(ContentType.Application.Json)
+                downloadProgress(listener)
```
Comment (e5l) [inline ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/ObservableBodyTest.kt:116]: We also could think about a better name here: the current name doesn't highlight the intention
- `listenProgress { }`
- `onDownload`
- `setProgressListener` or `progressListener`

Comment (rsinukov) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/ObservableBodyTest.kt:116]: i like `onDownload` to specify that it's the download and not upload progress

Comment (e5l) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/ObservableBodyTest.kt:116]: btw we also may want to make something like this for uploading

Comment (rsinukov) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/ObservableBodyTest.kt]: It's already merged #2450

Comment (rsinukov) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/ObservableBodyTest.kt]: Updated this PR to unify API

**Pull request merged**